packages/extension-ui/src/components/MnemonicInput.tsx

```diff
@@ -0,0 +1,168 @@
+// Copyright 2019-2023 @polkadot/extension-ui authors & contributors
+// SPDX-License-Identifier: Apache-2.0
+
+import type { ThemeProps } from '../types';
+
+import React, { useCallback, useEffect, useState } from 'react';
+import styled from 'styled-components';
+
+import { validateSeed } from '@polkadot/extension-ui/messaging';
+import { objectSpread } from '@polkadot/util';
+import { KeypairType } from '@polkadot/util-crypto/types';
+
+import { Warning } from '../components';
+import useTranslation from '../hooks/useTranslation';
+import { AccountInfo } from '../Popup/ImportSeed';
+import MnemonicPill from './MnemonicPill';
+
+interface Props extends ThemeProps {
+  className?: string;
+  onAccountChange: (account: AccountInfo | null) => void;
+  type: KeypairType;
+  seed: string | null;
+  genesis: string;
+  path: string | null;
+  address: string;
+  setAddress: (address: string) => void;
+  error: string;
+  setError: (error: string) => void;
+  onChange: (seed: string) => void;
+}
+
+const MnemonicInput = ({
+  address,
+  className,
+  error,
+  genesis,
+  onAccountChange,
+  onChange,
+  path,
+  seed,
+  setAddress,
+  setError,
+  type
+}: Props) => {
+  const [mnemonicWords, setMnemonicWords] = useState<string[]>(Array(12).fill(''));
+  const { t } = useTranslation();
+
+  const _handlePaste = useCallback(
+    (event: React.ClipboardEvent) => {
+      event.preventDefault();
+      const MNEMONIC_LENGTH = 12;
+
+      setMnemonicWords([]);
+      const pastedWords = event.clipboardData.getData('text').trim().split(' ').slice(0, 12);
+      const defaultValue = '';
+      const fillArray = Array(MNEMONIC_LENGTH - pastedWords.length).fill(defaultValue);
+      const result = pastedWords.concat(fillArray);
+
+      setMnemonicWords(result);
+      onChange(result.join(' '));
+    },
+    [onChange, setMnemonicWords]
+  );
+
+  const _handleChange = useCallback(
+    (value: string, index: number) => {
+      if (mnemonicWords.every((word) => word === '')) {
+        return;
+      }
+
+      const newMnemonicWords = [...mnemonicWords];
+
+      newMnemonicWords[index] = value;
+
+      onChange(newMnemonicWords.join(' '));
+      setMnemonicWords(newMnemonicWords.slice(0, 12));
+    },
+    [mnemonicWords, onChange, setMnemonicWords]
+  );
+
+  useEffect(() => {
+    setMnemonicWords(seed ? seed.split(' ') : Array(12).fill(''));
```

Comment:
If seed is not present shouldn't it just use the default empty mnemonicWords state instead of setting it up again? Or it's possible the seed would get change during lifecycle of the component and you want to reset to previous value? If the second is correct I would move `Array(12).fill('')` to a const outside component as initial state and use in both cases via spread